# Advanced Pandas Lab: Selection, Data Cleaning & GroupBy
**Course:** Introduction to Data Science  
**Topic:** Advanced Pandas  
**Software:** Python 3.x, Jupyter Notebook, Pandas  

---

## 1. Lab Overview & Objectives
This lab focuses on using Pandas for:
- Advanced row and column selection using `loc[]`, `iloc[]`, Boolean conditions, `isin()`, `between()`, and `query()`.
- Data cleaning techniques: identifying and handling missing values, identifying duplicate records, and standardizing inconsistent categorical values.
- Grouped analysis using `groupby()` with single and multiple aggregation functions, and sorting grouped results.
- End-to-end integrated data analysis workflow.


---
## 4. Dataset Preparation
We begin by creating `students.csv` with the specified dataset, which contains missing values, duplicate entries, and inconsistent department labels.


In [1]:
import pandas as pd
import numpy as np

csv_data = '''StudentID,Name,Age,Department,GPA,Fees
101,Ali,20,CS,3.20,85000
102,Sara,21,AI,3.70,90000
103,Ahmed,19,CS,2.80,82000
104,Hina,22,AI,3.50,90000
105,Usman,,SE,3.10,78000
106,Ayesha,23,cs,3.80,82000
107,Bilal,21,AI,,90000
108,Maria,24,SE,3.40,78000
109,Hamza,20,CS,3.00,85000
110,Zara,22,AI,3.90,90000
111,Ali,20,CS,3.20,85000
112,Noor,19,Computer Science,3.60,85000'''

with open("students.csv", "w") as f:
    f.write(csv_data)

print("students.csv successfully created!")


students.csv successfully created!


---
## 5. Setup and Initial Inspection
Load the dataset into a Pandas DataFrame and inspect its basic structure, shape, column names, data types, and non-null counts.


In [2]:
import pandas as pd

df = pd.read_csv("students.csv")

print("--- First 5 Rows (head) ---")
print(df.head())

print("\n--- DataFrame Shape ---")
print("Shape (Rows, Columns):", df.shape)

print("\n--- Column Names ---")
print(df.columns)

print("\n--- Data Types ---")
print(df.dtypes)

print("\n--- DataFrame Summary Info ---")
df.info()


--- First 5 Rows (head) ---
   StudentID   Name   Age Department  GPA   Fees
0        101    Ali  20.0         CS  3.2  85000
1        102   Sara  21.0         AI  3.7  90000
2        103  Ahmed  19.0         CS  2.8  82000
3        104   Hina  22.0         AI  3.5  90000
4        105  Usman   NaN         SE  3.1  78000

--- DataFrame Shape ---
Shape (Rows, Columns): (12, 6)

--- Column Names ---
Index(['StudentID', 'Name', 'Age', 'Department', 'GPA', 'Fees'], dtype='str')

--- Data Types ---
StudentID       int64
Name              str
Age           float64
Department        str
GPA           float64
Fees            int64
dtype: object

--- DataFrame Summary Info ---
<class 'pandas.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   StudentID   12 non-null     int64  
 1   Name        12 non-null     str    
 2   Age         11 non-null     float64
 3   Department  12 non-null

### Task 1 Answers
**Question:** Before changing anything, answer: How many rows and columns are present? Which columns contain numerical data? Which columns are categorical?

- **Number of Rows & Columns:** 
  - Total Rows: **12**
  - Total Columns: **6** (Shape: `(12, 6)`)
- **Numerical Columns:** 
  - `StudentID` (Integer identifier, `int64`)
  - `Age` (Numerical age with float representation due to missing NaN, `float64`)
  - `GPA` (Floating-point academic score, `float64`)
  - `Fees` (Numerical tuition fee, `int64`)
- **Categorical Columns:** 
  - `Name` (Textual name of student, `object` / `string`)
  - `Department` (Categorical department label, `object` / `string`)


---
## 6. Activity 1 — Advanced Selection and Filtering

### 6.1 Selecting Columns


In [3]:
# Task 2: Display only Name, Age, and Department
selected_cols = df[["Name", "Age", "Department"]]
print("Task 2 Output:")
print(selected_cols)


Task 2 Output:
      Name   Age        Department
0      Ali  20.0                CS
1     Sara  21.0                AI
2    Ahmed  19.0                CS
3     Hina  22.0                AI
4    Usman   NaN                SE
5   Ayesha  23.0                cs
6    Bilal  21.0                AI
7    Maria  24.0                SE
8    Hamza  20.0                CS
9     Zara  22.0                AI
10     Ali  20.0                CS
11    Noor  19.0  Computer Science


### 6.2 `loc[]` and `iloc[]`
- `loc[]`: Label-based indexing. Uses row index labels and column names.
- `iloc[]`: Integer position-based indexing. Uses 0-based positional indices.


In [4]:
# Task 3: Use loc[] to select the row whose index label is 4. Then use iloc[] to select the fifth row.

print("--- Using loc[4] (Index Label 4) ---")
print(df.loc[4])

print("\n--- Using iloc[4] (5th Row, Position Index 4) ---")
print(df.iloc[4])


--- Using loc[4] (Index Label 4) ---
StudentID       105
Name          Usman
Age             NaN
Department       SE
GPA             3.1
Fees          78000
Name: 4, dtype: object

--- Using iloc[4] (5th Row, Position Index 4) ---
StudentID       105
Name          Usman
Age             NaN
Department       SE
GPA             3.1
Fees          78000
Name: 4, dtype: object


### 6.3 Boolean Filtering


In [5]:
# Task 4: Find all students whose GPA is greater than 3.5
gpa_above_35 = df[df["GPA"] > 3.5]
print("--- Task 4: GPA > 3.5 ---")
print(gpa_above_35)


--- Task 4: GPA > 3.5 ---
    StudentID    Name   Age        Department  GPA   Fees
1         102    Sara  21.0                AI  3.7  90000
5         106  Ayesha  23.0                cs  3.8  82000
9         110    Zara  22.0                AI  3.9  90000
11        112    Noor  19.0  Computer Science  3.6  85000


In [6]:
# Task 5: Find students whose Age is at least 21
age_at_least_21 = df[df["Age"] >= 21]
print("--- Task 5: Age >= 21 ---")
print(age_at_least_21)


--- Task 5: Age >= 21 ---
   StudentID    Name   Age Department  GPA   Fees
1        102    Sara  21.0         AI  3.7  90000
3        104    Hina  22.0         AI  3.5  90000
5        106  Ayesha  23.0         cs  3.8  82000
6        107   Bilal  21.0         AI  NaN  90000
7        108   Maria  24.0         SE  3.4  78000
9        110    Zara  22.0         AI  3.9  90000


In [7]:
# Task 6: Find students who are in AI and have GPA greater than 3.5
ai_high_gpa = df[(df["Department"] == "AI") & (df["GPA"] > 3.5)]
print("--- Task 6: Department == 'AI' and GPA > 3.5 ---")
print(ai_high_gpa)


--- Task 6: Department == 'AI' and GPA > 3.5 ---
   StudentID  Name   Age Department  GPA   Fees
1        102  Sara  21.0         AI  3.7  90000
9        110  Zara  22.0         AI  3.9  90000


In [8]:
# Task 7: Find students from either AI or CS
ai_or_cs = df[(df["Department"] == "AI") | (df["Department"] == "CS")]
print("--- Task 7: Department is AI or CS ---")
print(ai_or_cs)


--- Task 7: Department is AI or CS ---
    StudentID   Name   Age Department  GPA   Fees
0         101    Ali  20.0         CS  3.2  85000
1         102   Sara  21.0         AI  3.7  90000
2         103  Ahmed  19.0         CS  2.8  82000
3         104   Hina  22.0         AI  3.5  90000
6         107  Bilal  21.0         AI  NaN  90000
8         109  Hamza  20.0         CS  3.0  85000
9         110   Zara  22.0         AI  3.9  90000
10        111    Ali  20.0         CS  3.2  85000


### 6.4 `isin()` and `between()`


In [9]:
# Task 8: Use isin() to select students from AI, CS, or SE
ai_cs_se = df[df["Department"].isin(["AI", "CS", "SE"])]
print("--- Task 8: Department in ['AI', 'CS', 'SE'] using isin() ---")
print(ai_cs_se)


--- Task 8: Department in ['AI', 'CS', 'SE'] using isin() ---
    StudentID   Name   Age Department  GPA   Fees
0         101    Ali  20.0         CS  3.2  85000
1         102   Sara  21.0         AI  3.7  90000
2         103  Ahmed  19.0         CS  2.8  82000
3         104   Hina  22.0         AI  3.5  90000
4         105  Usman   NaN         SE  3.1  78000
6         107  Bilal  21.0         AI  NaN  90000
7         108  Maria  24.0         SE  3.4  78000
8         109  Hamza  20.0         CS  3.0  85000
9         110   Zara  22.0         AI  3.9  90000
10        111    Ali  20.0         CS  3.2  85000


In [10]:
# Task 9: Use between() to select students aged 20 to 22
age_20_to_22 = df[df["Age"].between(20, 22)]
print("--- Task 9: Age between 20 and 22 using between() ---")
print(age_20_to_22)


--- Task 9: Age between 20 and 22 using between() ---
    StudentID   Name   Age Department  GPA   Fees
0         101    Ali  20.0         CS  3.2  85000
1         102   Sara  21.0         AI  3.7  90000
3         104   Hina  22.0         AI  3.5  90000
6         107  Bilal  21.0         AI  NaN  90000
8         109  Hamza  20.0         CS  3.0  85000
9         110   Zara  22.0         AI  3.9  90000
10        111    Ali  20.0         CS  3.2  85000


### 6.5 `query()`


In [11]:
# Task 10: Use query() to find students with GPA above 3.0 and Fees equal to 90000
query_result = df.query("GPA > 3.0 and Fees == 90000")
print("--- Task 10: query('GPA > 3.0 and Fees == 90000') ---")
print(query_result)


--- Task 10: query('GPA > 3.0 and Fees == 90000') ---
   StudentID  Name   Age Department  GPA   Fees
1        102  Sara  21.0         AI  3.7  90000
3        104  Hina  22.0         AI  3.5  90000
9        110  Zara  22.0         AI  3.9  90000


---
## 7. Activity 2 — Data Cleaning

### 7.1 Finding Missing Values


In [12]:
# Task 11: Identify missing values in each column
print("--- Task 11: Missing Value Counts per Column ---")
print(df.isna().sum())


--- Task 11: Missing Value Counts per Column ---
StudentID     0
Name          0
Age           1
Department    0
GPA           1
Fees          0
dtype: int64


### 7.2 Handling Missing GPA


In [13]:
# Task 12: Calculate mean GPA and fill missing GPA
mean_gpa = df["GPA"].mean()
print(f"Calculated Mean GPA: {mean_gpa:.4f}")
df["GPA"] = df["GPA"].fillna(mean_gpa)

# Task 13: Verify GPA no longer contains missing values
print("\nTask 13: Verification of missing values in GPA:", df["GPA"].isna().sum())


Calculated Mean GPA: 3.3818

Task 13: Verification of missing values in GPA: 0


### 7.3 Handling Missing Age


In [14]:
# Task 14: Fill missing Age using median age
median_age = df["Age"].median()
print(f"Calculated Median Age: {median_age:.1f}")
df["Age"] = df["Age"].fillna(median_age)

print("Verification of missing values in Age:", df["Age"].isna().sum())


Calculated Median Age: 21.0
Verification of missing values in Age: 0


**Question:** Why might median be preferable to mean when a variable contains extreme values?  
**Answer:** The median is a resistant measure of central location (it represents the middle value of sorted data) and is not sensitive to extreme outliers. Conversely, the arithmetic mean incorporates every value into its calculation, so extreme low or high values (outliers) heavily skew the mean, making it unrepresentative of the typical value in skewed distributions.


### 7.4 Finding Duplicates


In [15]:
# Task 15: Identify and remove duplicate records
print("Duplicates checking all columns (including StudentID):", df.duplicated().sum())

# Note: Student 111 (Ali) has duplicate student details as Student 101 except for StudentID.
# Checking duplicate records ignoring StudentID:
subset_cols = ["Name", "Age", "Department", "GPA", "Fees"]
print("Duplicate student records (ignoring StudentID):", df.duplicated(subset=subset_cols).sum())

# Remove duplicate records
df = df.drop_duplicates(subset=subset_cols).reset_index(drop=True)
print("Shape after removing duplicate student records:", df.shape)
print("\nCleaned DataFrame after duplicate removal:")
print(df)


Duplicates checking all columns (including StudentID): 0
Duplicate student records (ignoring StudentID): 1
Shape after removing duplicate student records: (11, 6)

Cleaned DataFrame after duplicate removal:
    StudentID    Name   Age        Department       GPA   Fees
0         101     Ali  20.0                CS  3.200000  85000
1         102    Sara  21.0                AI  3.700000  90000
2         103   Ahmed  19.0                CS  2.800000  82000
3         104    Hina  22.0                AI  3.500000  90000
4         105   Usman  21.0                SE  3.100000  78000
5         106  Ayesha  23.0                cs  3.800000  82000
6         107   Bilal  21.0                AI  3.381818  90000
7         108   Maria  24.0                SE  3.400000  78000
8         109   Hamza  20.0                CS  3.000000  85000
9         110    Zara  22.0                AI  3.900000  90000
10        112    Noor  19.0  Computer Science  3.600000  85000


### 7.5 Checking Category Consistency


In [16]:
print("--- Unique Department values before cleaning ---")
print(df["Department"].unique())

print("\n--- Value counts before cleaning ---")
print(df["Department"].value_counts())

# Task 16: Standardize 'cs' as 'CS' and 'Computer Science' as 'CS'
df["Department"] = df["Department"].replace({
    "cs": "CS",
    "Computer Science": "CS"
})

# Task 17: Run value_counts() again to verify
print("\n--- Task 17: Department Value Counts After Standardization ---")
print(df["Department"].value_counts())


--- Unique Department values before cleaning ---
<StringArray>
['CS', 'AI', 'SE', 'cs', 'Computer Science']
Length: 5, dtype: str

--- Value counts before cleaning ---
Department
AI                  4
CS                  3
SE                  2
cs                  1
Computer Science    1
Name: count, dtype: int64

--- Task 17: Department Value Counts After Standardization ---
Department
CS    5
AI    4
SE    2
Name: count, dtype: int64


---
## 8. Activity 3 — Grouping and Aggregation

### 8.1 Basic GroupBy


In [17]:
# Task 18: Calculate average GPA for each department
avg_gpa_dept = df.groupby("Department")["GPA"].mean()
print("--- Task 18: Average GPA by Department ---")
print(avg_gpa_dept)


--- Task 18: Average GPA by Department ---
Department
AI    3.620455
CS    3.280000
SE    3.250000
Name: GPA, dtype: float64


In [18]:
# Task 19: Calculate the number of students in each department
student_count_dept = df.groupby("Department")["StudentID"].count()
print("--- Task 19: Student Count by Department ---")
print(student_count_dept)


--- Task 19: Student Count by Department ---
Department
AI    4
CS    5
SE    2
Name: StudentID, dtype: int64


In [19]:
# Task 20: Calculate total fees collected from each department
total_fees_dept = df.groupby("Department")["Fees"].sum()
print("--- Task 20: Total Fees Collected by Department ---")
print(total_fees_dept)


--- Task 20: Total Fees Collected by Department ---
Department
AI    360000
CS    419000
SE    156000
Name: Fees, dtype: int64


### 8.2 Multiple Aggregations


In [20]:
# Task 21: For each department, calculate mean, min, and max GPA
gpa_agg = df.groupby("Department")["GPA"].agg(["mean", "min", "max"])
print("--- Task 21: GPA Aggregations (mean, min, max) by Department ---")
print(gpa_agg)


--- Task 21: GPA Aggregations (mean, min, max) by Department ---
                mean       min  max
Department                         
AI          3.620455  3.381818  3.9
CS          3.280000  2.800000  3.8
SE          3.250000  3.100000  3.4


In [21]:
# Task 22: For each department, calculate average Age and average GPA
age_gpa_avg = df.groupby("Department")[["Age", "GPA"]].mean()
print("--- Task 22: Average Age and Average GPA by Department ---")
print(age_gpa_avg)


--- Task 22: Average Age and Average GPA by Department ---
             Age       GPA
Department                
AI          21.5  3.620455
CS          20.2  3.280000
SE          22.5  3.250000


### 8.3 Sorting Grouped Results


In [22]:
# Task 23: Sort departments by average GPA from highest to lowest
sorted_gpa_dept = df.groupby("Department")["GPA"].mean().sort_values(ascending=False)
print("--- Task 23: Departments Sorted by Average GPA (Descending) ---")
print(sorted_gpa_dept)


--- Task 23: Departments Sorted by Average GPA (Descending) ---
Department
AI    3.620455
CS    3.280000
SE    3.250000
Name: GPA, dtype: float64


---
## 9. Integrated Challenge — Clean and Analyze
In this section, we re-execute the complete end-to-end data processing workflow from a raw loading step to final summary stats and write key conclusions.


In [23]:
# Step 1: Load raw data
raw_df = pd.read_csv("students.csv")

# Step 2: Inspect raw data
print("Raw Shape:", raw_df.shape)
print("Missing values per column:\n", raw_df.isna().sum())

# Step 3: Remove duplicate records
cleaned_df = raw_df.drop_duplicates(subset=["Name", "Age", "Department", "GPA", "Fees"]).copy()

# Step 4: Fill missing Age using median
median_age = cleaned_df["Age"].median()
cleaned_df["Age"] = cleaned_df["Age"].fillna(median_age)

# Step 5: Fill missing GPA using mean
mean_gpa = cleaned_df["GPA"].mean()
cleaned_df["GPA"] = cleaned_df["GPA"].fillna(mean_gpa)

# Step 6: Standardize department names
cleaned_df["Department"] = cleaned_df["Department"].replace({
    "cs": "CS",
    "Computer Science": "CS"
})

# Step 7: Find students with GPA >= 3.5
high_achievers = cleaned_df[cleaned_df["GPA"] >= 3.5]
print("\n--- Students with GPA >= 3.5 ---")
print(high_achievers[["StudentID", "Name", "Department", "GPA"]])

# Step 8: Calculate student count by department
dept_counts = cleaned_df.groupby("Department")["StudentID"].count()

# Step 9: Calculate average GPA by department
dept_avg_gpa = cleaned_df.groupby("Department")["GPA"].mean()

# Step 10: Calculate total fees by department
dept_total_fees = cleaned_df.groupby("Department")["Fees"].sum()

# Step 11: Combine and Sort departments by average GPA
dept_summary = pd.DataFrame({
    "Student_Count": dept_counts,
    "Average_GPA": dept_avg_gpa,
    "Total_Fees": dept_total_fees
}).sort_values(by="Average_GPA", ascending=False)

print("\n--- Final Department Summary Table (Sorted by Average GPA) ---")
print(dept_summary)


Raw Shape: (12, 6)
Missing values per column:
 StudentID     0
Name          0
Age           1
Department    0
GPA           1
Fees          0
dtype: int64

--- Students with GPA >= 3.5 ---
    StudentID    Name Department  GPA
1         102    Sara         AI  3.7
3         104    Hina         AI  3.5
5         106  Ayesha         CS  3.8
9         110    Zara         AI  3.9
11        112    Noor         CS  3.6

--- Final Department Summary Table (Sorted by Average GPA) ---
            Student_Count  Average_GPA  Total_Fees
Department                                        
AI                      4        3.625      360000
CS                      5        3.280      419000
SE                      2        3.250      156000


### Integrated Challenge Observations & Conclusions
Based on the cleaned dataset analysis:

1. **AI Department Academic Performance:**  
   The **AI (Artificial Intelligence)** department achieved the highest academic performance overall with an average GPA of **3.6205**. Additionally, AI accounts for 3 of the 5 students who earned a GPA of 3.5 or higher.

2. **CS Department Revenue & Enrollment Dominance:**  
   The **CS (Computer Science)** department has the highest enrollment (**5 students**, 45.5% of total clean student body) and generates the largest tuition fee revenue (**8419,000**), despite having a slightly lower average GPA (**3.2800**) than AI.

3. **SE Department Demographics:**  
   The **SE (Software Engineering)** department has the smallest student count (**2 students**), the highest average age (**22.5 years**), and the lowest total revenue (**\$156,000**), with an average GPA of **3.2500**.

4. **Impact of Data Standardization and Cleaning:**  
   Before cleaning, category inconsistencies (`cs`, `Computer Science`, `CS`) fragmented the Computer Science cohort across three labels. Standardizing labels and filling missing values enabled accurate aggregation and prevented revenue and GPA under-reporting for CS.


---
## 10. Optional Extension


In [24]:
# Calculate department average GPA and transform back to original shape
cleaned_df["Department_Avg_GPA"] = cleaned_df.groupby("Department")["GPA"].transform("mean")

# Create Above_Department_Average boolean column
cleaned_df["Above_Department_Average"] = cleaned_df["GPA"] > cleaned_df["Department_Avg_GPA"]

print("--- DataFrame with Department Average & Above Average Indicator ---")
print(cleaned_df[["StudentID", "Name", "Department", "GPA", "Department_Avg_GPA", "Above_Department_Average"]])


--- DataFrame with Department Average & Above Average Indicator ---
    StudentID    Name  ... Department_Avg_GPA  Above_Department_Average
0         101     Ali  ...              3.280                     False
1         102    Sara  ...              3.625                      True
2         103   Ahmed  ...              3.280                     False
3         104    Hina  ...              3.625                     False
4         105   Usman  ...              3.250                     False
5         106  Ayesha  ...              3.280                      True
6         107   Bilal  ...              3.625                     False
7         108   Maria  ...              3.250                      True
8         109   Hamza  ...              3.280                     False
9         110    Zara  ...              3.625                      True
11        112    Noor  ...              3.280                      True

[11 rows x 6 columns]


---
## 11. Expected Skills Checklist

| Skill Description | Status |
| :--- | :---: |
| Select one or multiple columns | [x] Completed |
| Use `loc[]` | [x] Completed |
| Use `iloc[]` | [x] Completed |
| Filter using Boolean conditions | [x] Completed |
| Use `isin()` | [x] Completed |
| Use `between()` | [x] Completed |
| Use `query()` | [x] Completed |
| Detect missing values | [x] Completed |
| Fill missing values | [x] Completed |
| Detect and remove duplicates | [x] Completed |
| Inspect unique categorical values | [x] Completed |
| Use `value_counts()` | [x] Completed |
| Use `groupby()` | [x] Completed |
| Apply aggregation functions | [x] Completed |
| Sort grouped results | [x] Completed |
| Interpret grouped statistics | [x] Completed |


---
## 12. Short Viva / Discussion Questions

### Question 1: What is the difference between `loc[]` and `iloc[]`?
**Answer:**
- `loc[]` is **label-based**. It accesses rows and columns using explicit index labels (e.g., index label `4`, `'Department'`) and includes the stop boundary when slicing (`0:4` includes label `4`).
- `iloc[]` is **integer position-based**. It accesses elements using 0-indexed positions (e.g., row `0`, column `2`) regardless of label names, and excludes the stop boundary when slicing (`0:4` includes positions `0, 1, 2, 3`).

---

### Question 2: Why can `loc[2]` and `iloc[2]` return different rows?
**Answer:**
`loc[2]` looks for a row whose **index label** is literally `2`. If the DataFrame's index has been re-indexed, filtered, sorted, or set to custom strings (e.g. `['a', 'b', 'c']`), index label `2` might refer to a non-third row or might not exist at all.  
`iloc[2]` always refers strictly to the **3rd physical row** (0-indexed position 2) in the DataFrame, regardless of index label names.

---

### Question 3: What is the difference between `dropna()` and `fillna()`?
**Answer:**
- `dropna()` **removes/deletes** rows or columns containing missing values (`NaN` / `None`) from the DataFrame.
- `fillna()` **replaces/imputes** missing values (`NaN`) with a specified constant value, summary statistic (e.g., mean, median, mode), or propagated values (`ffill`/`bfill`), keeping all rows intact.

---

### Question 4: Why should duplicate records be checked before analysis?
**Answer:**
Duplicate records introduce artificial data inflation. They bias summary statistics (e.g., artificially skewing means, counts, and sums), distort machine learning training distributions, cause overcounting in financial metrics (like total fees), and lead to inaccurate business conclusions. Checking and dropping duplicates ensures data integrity.

---

### Question 5: What does `value_counts()` tell us?
**Answer:**
`value_counts()` returns a Pandas Series containing the frequency counts of unique values present in a Series or DataFrame column, sorted in descending order by default. It helps identify common categories, rare values, dominant classes, and inconsistent categorical entries (e.g., `cs` vs `CS`).

---

### Question 6: What does `groupby()` do?
**Answer:**
`groupby()` implements the **Split-Apply-Combine** strategy:
1. **Split:** Divides the DataFrame into subset groups based on unique values in specified key columns.
2. **Apply:** Applies an aggregation, transformation, or filtering function independently to each group.
3. **Combine:** Merges the group results into a single output data structure (Series or DataFrame).

---

### Question 7: What is the difference between `mean()`, `count()`, and `sum()` after `groupby()`?
**Answer:**
- `mean()`: Computes the arithmetic average of numeric values within each group.
- `count()`: Counts the number of non-null values present within each group.
- `sum()`: Calculates the total numerical sum of values within each group.

---

### Question 8: Why is data cleaning performed before final analysis?
**Answer:**
Data cleaning ensures the accuracy, quality, and reliability of analytical results ("Garbage In, Garbage Out"). Raw data often contains missing values, duplicates, formatting inconsistencies, and invalid inputs. Performing analysis on uncleaned data leads to invalid group counts, distorted averages, broken queries, and flawed decision-making.
